### DISCUSSION TITLE
Moving libraries and files between locations

### DISCUSSION THREAD

**Discussion by timcorbin** [latenitefilms/BRAWToolbox#88] [Support] [answered]
This is related to the discussion about media management. I'm trying to figure out the issues we've been having. Here's the current workflow used by myself and one other editor:

- We each have a 14" 2021 MacBook Pro M1 Max w/ 32GB RAM. We carry that with us and each only edit from our machine both at the office and at home.
- At the office we both edit from a single SMB NAS (IP over Thunderbolt; a QNAP TVS-872XT)
- At home: We each have a portable DAS that's fully mirrored from the NAS before we go home
- Media and cache are stored in a folder, not the library, but the entire project (all media, cache, library, etc.) are all contained in a folder structure for that editing project

This has never been an issue until we started using BRAW Toolbox. Now, on the DAS at home when we open a library that has files imported from BRAW Toolbox the machine keeps searching for the SMB NAS, as if that's where it thinks the media is. Final Cut will also tend to beachball here. A couple times I've been able to relink the media to the DAS (our standard workflow) but that changes nothing. It feels like the BRAW Toolbox media isn't being relinked to the DAS path.

Is there a better way to do this or a fix? Not being able to relink media makes Toolbox compeletely unusable as part of the workflow. 

Thanks!

**Comment by latenitefilms** [comment, marked as answer]
Do your NAS/DAS both have the same file path, or does this change? Do you have to manually relink the other non-BRAW Toolbox clips, or do they automagically relink?

Our preferred collaboration workflow is discussed [here](https://brawtoolbox.io/collaboration/). Essentially, if the file path remains the same, everything should just work.

However, if the file path changes, there is a [Toolbox](https://brawtoolbox.io/toolbox/#relink-braw-clips-within-an-event) that allows you to re-link files. Personally though, I would probably try and avoid this, as it would be annoying having to manually relink files as you constantly move between locations.

If your NAS and DAS have different paths, one workaround would be to use a sym-link to "trick" your system into thinking that files are stored at the same path. For example, if at work your media is stored in `/Volumes/NAS/MyMedia`, but at home your media is stored in `/Volumes/DAS/Work/MyMedia`, you could create a sym-link so that `/Volumes/NAS/MyMedia` points to `/Volumes/DAS/Work/MyMedia`.

**Reply by timcorbin** [reply]
The DAS and NAS do have different paths. Symlinks won't work at home if the NAS not present, right? At least in my testing they don't. It's a NAS, but we use it more like a DAS. It's not connected to the internet, so when we're at home, they only versions we have of anything are on the DAS. 

I'll give the Toolbox relinking a try and see if that works.

**Reply by latenitefilms** [reply]
Symlinks at home should work. You basically create a "fake NAS" path that points to your DAS path, so that as far as FCPX is concerned, the paths remain the same. This has the benefit that you shouldn't have to do any relinking for any media.

**Reply by timcorbin** [reply]
Maybe I'm not understanding symlinks, but how can they work if the target is offline? When at home, the NAS is not accessible. A symlink on the DAS points to its target on the NAS. If the NAS is not connected, what can the symlink access? Also, the DAS needs to have new data written to it when at home. New/changed data is then synced back to the NAS when we're in the office. Then we edit off the NAS while there, sync back to our respective DAS, and each take our DAS home. When I test this at the office, as soon as I disconnect the NAS, the symlinks go offline. 

I'm not sure how else to manage media for this workflow other than the way we've been doing it. A full copy of the media and library has to exist on the DAS in order for it be accessible. 

I tried relinking today from within Toolbox and it just beach balled FCP; I'll try again tomorrow. I haven't been able to get this popup window yet when relinking. I'm obviously missing something.
<img width="674" alt="image" src="https://user-images.githubusercontent.com/17440099/220507745-7749aff7-3c48-4fb9-8bd2-84c0d4e3f88f.png">

**Reply by latenitefilms** [reply]
Lets say for instance, at work, your NAS mounts to `/Volumes/NAS`. You might have a folder called `Jobs` on the NAS and within that you have a individual job folder and then maybe within that you have a `Media` folder. So at work, when you import a BRAW clip, it points to something like: `/Volumes/NAS/Jobs/JobA/Media/ClipA.braw`.

In the Final Cut Pro library, that absolute path is stored internally - i.e. `/Volumes/NAS/Jobs/JobA/Media/ClipA.braw` is what you will see in the FCPXML.

Now, when you go home, obviously `/Volumes/NAS/Jobs/JobA/Media/ClipA.braw` doesn't actually exist. Instead, you have `ClipA.braw` stored on a completely different drive, let's say, for example `/Volumes/DAS/Jobs/JobA/Media/ClipA.braw`.

In this simple example, the issue is that at work you have everything on a volume called `NAS`, but at home you have everything on a volume called `DAS`.

We can use sym-links to solve this, so that at home, when Final Cut Pro looks as the `NAS` volume - it's ACTUALLY pointing to the `DAS` volume.

Essentially you're "redirecting" `/Volumes/NAS/` to `/Volumes/DAS`.

You can use the `ln` Terminal command to create a sym-link, for example:

```
ln -s /path/to/original /path/to/link
```

You can read more [here](https://www.howtogeek.com/297721/how-to-create-and-use-symbolic-links-aka-symlinks-on-a-mac/).

Basically, you use sym-links to "trick" macOS and Final Cut Pro into thinking the NAS is connected, when the files are ACTUALLY stored on your DAS.

You will also have to allow sandbox access to your DAS path - see: https://brawtoolbox.io/collaboration/

To use the **Relink BRAW Clips within an EVENT** Toolbox, you simply drag an EVENT which contains the BRAW clips into the Toolbox drop zone.

For example, I have this test event:

<img width="965" alt="image" src="https://user-images.githubusercontent.com/22286696/220543548-86f418b5-baae-4c3f-892e-94ea2415ab8b.png">

...which I drag to here:

<img width="888" alt="image" src="https://user-images.githubusercontent.com/22286696/220543622-43fe2cf9-fbcb-4542-b914-cb0be29affdb.png">

I then get presented with this:

<img width="982" alt="image" src="https://user-images.githubusercontent.com/22286696/220543685-af548852-1889-4da1-b0bb-9e76ec0a7b98.png">

In this case all my media is "online" so it's all happy green dots. However, if it was red or orange dots, I could manually correct the path, then press **Refresh & Request Permissions** to update things. Once they're all green again, I press the **Relink BRAW Files** buttons to generate a new FCPXML, then drag the event icon back into my Final Cut Pro library.

However, as explained, this Toolbox was really only designed as a "worse case scenario" - and it would be much easier/better if you just kept the media in the same path to avoid any re-linking hell.

You might also be able to avoid using symlinks if you can just rename your DAS drives to the same drive name as your NAS.

Feel free to share a FCPXML or FCPX Library from both home and work if you want me to give you specific examples to your exact paths.

Hope this helps!

**Reply by timcorbin** [reply]
Chris, I super appreciate you sharing all this and I will work through it more today, and I think I understand the concept of what symlinks do but I really just don't see how they will actually work. 

The symlink points to the media on the `NAS`, which isn't present at home. How is Final Cut going to find the media if it isn't there? I've tested this, it doesn't work. I am hoping based on your confidence on this working that I'm doing something wrong, but I've created the symlinks just like your example (and the hundred others I've come across in trying to solve this) and it always is just referencing the media that exists elsewhere, and if that media no longer is present, the symlink breaks.

Let's say I create a symlink on the `DAS` to the `NAS` project folder, so symlink for `/Volumes/DAS/Live/` and pointing to `/Volumes/NAS/Live/`. That `Live` project folder has 15-20 active projects inside (each with their own folder structure). The symlink will work fine but it's now pointing to the media on the `NAS`, not the `DAS`, and it works fine UNTIL the `NAS` is not present, then the symlink breaks. How can the media that's already on the `DAS` be made to look like it's on the `NAS` if Final Cut can't see the symlink because it's broken?

At this point I may try renaming the drives so they're all the same. I'll try on Friday when I'm in the office again.

**Reply by timcorbin** [reply]
I did get the media on the `DAS` at home to work by just editing the BRAW paths in an XML and reimporting it — but that's obviously not a sustainable workflow. The Toolbox relinking says no Toolbox media was found. Do I have to keep the Final Cut Event that Toolbox creates or can I just move that Event into the main Event to merge the BRAW footage with other content (that's what I've been doing and it works, except for the relinking part)? I don't really want to have multiple events, but if that's the only way, then ok. Thanks again.

**Reply by latenitefilms** [reply]
I'm on a shoot this morning, but I'll try record a demonstration video explaining what I mean later today or tomorrow.

**Reply by timcorbin** [reply]
I just tried renaming our `NAS` and two `DAS` to have the same name. Seems to be working so far. This will indeed be WAY easier than relinking. Will update if anything changes. :)

**Reply by latenitefilms** [reply]
Great news - this is definitely way easier than setting up symlinks!

**Reply by timcorbin** [reply]
Well, I thought we were good to go. However, I've just opened up a library at home on the `DAS` that has all the paths correctly named — I checked an XML I was able to export — and it's now still searching for the `NAS` with a popup every 20 seconds or so saying it can't find the `NAS`. But all the BRAW media is loading correctly in the timeline (this wouldn't happen before — it'd be offline). FCP keeps beach balling before each popup, making it unusable. Even when I'm able to get the event dragged into Toolbox to relink, the files never load up in Toolbox (although I can't then drag the event in again). If close Toolbox after this and then try to reopen it, it just stays on the 'Loading' screen.

This is now a major issue, as I have a deadline. 

Any suggestions?

**Reply by timcorbin** [reply]
Just reporting stuff as it happens: Created a new library and pasted the timeline into a new project. Closed the original library. For some reason, that seems to have worked so far. 

When you have time Chris, let me know what you think might be going on. I'll report back if anything changes.

**Reply by latenitefilms** [reply]
Your previous message is a bit confusing. If you grant BRAW Toolbox Sandbox Access to your hard drive, basically BRAW Toolbox will internally only use the path as it appears in the FCPX Inspector within the BRAW Toolbox effect. So if the path is right, it'll work.

The Relink features in FCPX have no knowledge of BRAW Toolbox, so you can't use them to relink BRAW Clips.

If something is popping up wanting the NAS, it sounds like you have some other clips in FCPX that might still be pointing to it?

If you can share some screenshots or the FCPXML that might help work out what you're seeing.

**Reply by timcorbin** [reply]
Oh man, I think it was the Sandbox Access. I thought I had done that, but obviously not. As soon as I granted it, the errors went away. 

So, now that I have my tail between my legs, I'll just ask: Is there anyway to make granting that access part of the import process in Toolbox — a checkbox that grants access without having to then go into the clips? I know this sounds lazy, but it always nice to have a single-step import process. 

Anyways, thanks for bearing with me — and thanks again for all your help. I'll pop back in if anything changes.

**Reply by latenitefilms** [reply]
Unfortunately not - the Workflow Extension and the FxPlug4 Effect within Final Cut Pro are technically two seperate applications, so currently when you import clips in the Workflow Extension we create a security-scope bookmark that the FxPlug4 effect can resolve, giving us sandbox access to the clip, but this only works on an individual machine.

In your case though, if you're always using the same DAS, you should only have to grant sandbox access to your external drive once.

Untested, and unlikely to work, but MAYBE if you copy this file from your work machine to your home machine, it MIGHT work if the file paths remain the same between machines. I'll look into this myself at some point this week.

`~/Library/Group Containers/A5HDJTY9X5.com.latenitefilms.BRAWToolbox/Library/Application Support/BRAWToolbox.document`

Unfairly, despite sandboxing being a requirement of everything on the Mac App Store, Apple's own Final Cut Pro isn't actually sandboxed, so they get around a lot of the sandbox headaches.
